File: backend/app/services/matching.py
from dataclasses import dataclass
from enum import StrEnum

from sqlalchemy import ColumnElement, func
from sqlalchemy.orm import InstrumentedAttribute


class MatchStatus(StrEnum):
    NONE = "none"
    UNIQUE = "unique"
    AMBIGUOUS = "ambiguous"


@dataclass(frozen=True, slots=True)
class MatchCandidate:
    id: int
    name: str
    matched_patterns: tuple[str, ...]


@dataclass(frozen=True, slots=True)
class RuleMatch:
    status: MatchStatus
    candidates: tuple[MatchCandidate, ...]


def patterns_match_clause(
    recipient: ColumnElement[str] | InstrumentedAttribute[str],
    subject: ColumnElement[str] | InstrumentedAttribute[str],
    patterns: ColumnElement[str] | InstrumentedAttribute[str] | str,
) -> ColumnElement[bool]:
    return func.finances_patterns_match(recipient, subject, patterns) == 1


def rule_match(candidates: tuple[MatchCandidate, ...]) -> RuleMatch:
    if not candidates:
        status = MatchStatus.NONE
    elif len(candidates) == 1:
        status = MatchStatus.UNIQUE
    else:
        status = MatchStatus.AMBIGUOUS
    return RuleMatch(status=status, candidates=candidates)

File: frontend/src/types/matching.ts
import * as z from 'zod'

const MatchCandidateSchema = z
  .object({
    id: z.number().int().positive(),
    name: z.string(),
    matched_patterns: z.array(z.string())
  })
  .strict()

export const RuleMatchSchema = z
  .object({
    status: z.enum(['none', 'unique', 'ambiguous']),
    candidates: z.array(MatchCandidateSchema)
  })
  .strict()

export type RuleMatch = z.infer<typeof RuleMatchSchema>
